### PR TITLE
Implement streaming completion

### LINKED ISSUE #7: Add streaming support
Opening an issue to track this feature, since it seems to keep coming up, and I'm expecting to make use of it soon for a project.

See also:
- https://www.beskar.co/blog/streaming-openai-completions-vercel-edge
- https://www.npmjs.com/package/@beskar-labs/parse-json-sse
- https://github.com/transitive-bullshit/chatgpt-api/blob/main/src/fetch-sse.ts

### PR DESCRIPTION
**Pull request by dglazkov**
Here's a streaming completion implementation. I tried to stick to the spirit of the codebase.

PTAL.

Fixes #7.

### CODE REVIEW

**Comment by dglazkov** [comment]
Here's a working example: https://github.com/dglazkov/openai-fetch-fun

**Comment by rileytomasek** [comment]
Thanks @dglazkov — this is awesome! I will work on reviewing it ASAP.

File: src/streaming.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        const pos = line.indexOf(':');
+        const name = line.substring(0, pos);
+        if (name !== 'data') return;
+        const content = line.substring(pos + 1).trim();
+        if (content.length == 0) return;
+        if (content === '[DONE]') {
+          this.onend?.();
+          return;
+        }
+        const parsed = JSON.parse(content);
```
Comment (transitive-bullshit) [inline src/streaming.ts]: def add a try / catch around `JSON.parse`

I've seen examples in the wild with https://github.com/transitive-bullshit/chatgpt-api where this unexpectedly fails

File: src/streaming.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        this.onchunk?.({
+          completion: parsed.choices[0].text || '',
```
Comment (transitive-bullshit) [inline src/streaming.ts]: I think you should just pass the parsed content directly instead of adding additional metadata

this.onchunk?.(parsed)

**Review by transitive-bullshit — CHANGES_REQUESTED** [review]
this looks really awesome @dglazkov 🔥 

My main question would be: how does this work from Vercel edge functions (and edge functions in general)? Where you want to stream the response from openai but you don't want to process it within the edge function handler itself.

I don't think the current solution handles that, and it's an important use case for a lot of apps where `openai-fetch` comes into play. Other apps may use `openai` or the REST API directly, but `openai-fetch` sweet spot is in serverless and edge use cases (at least for now). So just wanna make sure that the streaming implementation really works well with that.

really really great work on this, though. and we may want to just merge as is @rileytomasek since imho it's already a huge improvement.

File: src/streaming.ts (hunk 1 above)
Comment (dglazkov) [reply src/streaming.ts]: Done!

File: src/streaming.ts (hunk 2 above)
Comment (dglazkov) [reply src/streaming.ts]: I wanted to make sure this stays in spirit with the rest of the API, so I kept the shape of the object the same as the response in non-streaming completion.

**Comment by dglazkov** [comment]
@transitive-bullshit this code will work in Node 18+, Web main thread and Web workers. The latter includes most edge workers that I know of and deno.

File: src/streaming.ts (hunk 2 above)
Comment (rileytomasek) [reply src/streaming.ts]: Yeah, I think it's best to keep it consistent with the rest of the API.

**Comment by rileytomasek** [comment]
Do you mind adding `streamChatCompletion()` for completeness? It may be confusing for one to support streaming, but not the other.

The Vercel edge functions and app directory have a lot of bugs with standards compliant implementations because they are modifying fetch for caching and other things. Did you (or can you) create a simple test to make sure the streaming works with them? As @transitive-bullshit said, it's not the end of the world if not because this is a huge improvement as is and Vercel support can be added later since the API is unlikely to change because it should just be implementation changes.

File: src/fetch-api.ts [hunk 3]
```diff
@@ -4,7 +4,7 @@ import { OpenAIApiError } from './errors';
 
 const DEFAULT_BASE_URL = 'https://api.openai.com/v1';
 
-export interface FetchOptions extends Options {
+export interface FetchOptions extends Omit<Options, 'credentials'> {
```
Comment (rileytomasek) [inline src/fetch-api.ts:7]: Thanks for catching this

**Comment by dglazkov** [comment]
I'll add the `streamChatCompletion`. I do want to try it with the edge functions, but I haven't gotten to that yet.

**Comment by transitive-bullshit** [comment]
Excellent work @dglazkov 🔥

**Comment by dglazkov** [comment]
Added streaming chat completion support.

**Comment by ralyodio** [comment]
when merge?

**Comment by rileytomasek** [comment]
🥳 Thank you very much @dglazkov! I'm going to release a new version with this now.

**Comment by rileytomasek** [comment]
It has been released as v1.2.0: https://github.com/rileytomasek/openai-fetch/releases/tag/v1.2.0

cc @ralyodio

**Pull request merged**